Allow dnn convolution to work inplace

abergeron opened the pull request:
Add alpha param and an inplace option.  'alpha' is not supported in cudnn v1 so we disable optimizations that would set it to something other than 1 with that version.

This may be missing a test for the merging optimizations.

lamblin commented on theano/sandbox/cuda/dnn.py:
```diff
@@ -355,88 +374,105 @@ def __init__(self, workmem=None):
         self.__dict__.update(d)
         if not hasattr(self, 'workmem'):
-            self.workmem = 'small'
+            self.workmem = 'none'
```
Why change the default here?

abergeron replied:
Because this way, it has the same behaviour as before, and we can avoid testing for the cases that fail there.

lamblin replied:
OK.

lamblin commented on theano/sandbox/cuda/dnn.py:
```diff
@@ -355,88 +374,105 @@ def __init__(self, workmem=None):
 
-        return d_img, d_kerns, theano.gradient.DisconnectedType()()
+        return [d_img, d_kerns, output.zeros_like(),
+                DisconnectedType()(), DisconnectedType()()]
```
I don't think the gradient wrt `output` and `alpha` are correct.
If I understood correctly, the gradient wrt `output` should be something like `alpha * top`, and the gradient wrt `alpha` should be `(top * output).sum()` or something.

lamblin replied:
If you cannot implement that now, you can use `grad_not_implemented` instead.

abergeron replied:
I'd rather have the proper gradient if it's not too hard.  I'll make some tests for the grad to make sure.

abergeron replied:
I fixed the gradient for output and changed alpha to grad_not_implemented since I couldn't find a expression that works.  I don't think that too many people will actually use the gradient on alpha so it should be ok.

lamblin replied:
alpha just scales the output of the convolution, is that right?
If I understand correctly now, contrary to what I suggested, it is not the original `output` that should be used, but the output of the convolution part only (the thing that is multiplied by alpha).
So, something like `(top * GpuDnnConv()(img, kerns, output.zeros_like(), desc).sum()`.
It may not be practical to use, so I'm ok with not_implemented.

abergeron replied:
I've tried with that and the gradient was still wrong.  Which is why I just left it not implemented.

lamblin commented on theano/sandbox/cuda/dnn.py:
```diff
@@ -355,88 +374,105 @@ def __init__(self, workmem=None):
-        # not connected to desc
-        return [[1], [1], [0]]
+        # not connected to desc, alpha
+        return [[1], [1], [1], [0], [0]]
```
Maybe I don't get what alpha is doing, then, but doesn't it change the output?

abergeron replied:
Yep, alpha is not connected.

abergeron commented:
I've fixed the gradients and added a test for them.  I've also fixed the last problems in the merge opts and the test for them.  This should now be ready for merge.
